cpp/src/arrow/util/future.cc

```diff
@@ -250,30 +250,40 @@ class ConcreteFutureImpl : public FutureImpl {
   }
 
   void DoMarkFinishedOrFailed(FutureState state) {
-    {
-      // Lock the hypothetical waiter first, and the future after.
-      // This matches the locking order done in FutureWaiter constructor.
-      std::unique_lock<std::mutex> waiter_lock(global_waiter_mutex);
-      std::unique_lock<std::mutex> lock(mutex_);
-
-      DCHECK(!IsFutureFinished(state_)) << "Future already marked finished";
-      state_ = state;
-      if (waiter_ != nullptr) {
-        waiter_->MarkFutureFinishedUnlocked(waiter_arg_, state);
-      }
+    // Lock the hypothetical waiter first, and the future after.
+    // This matches the locking order done in FutureWaiter constructor.
+    std::unique_lock<std::mutex> waiter_lock(global_waiter_mutex);
+    std::unique_lock<std::mutex> lock(mutex_);
+
+    DCHECK(!IsFutureFinished(state_)) << "Future already marked finished";
+    state_ = state;
+    if (waiter_ != nullptr) {
+      waiter_->MarkFutureFinishedUnlocked(waiter_arg_, state);
     }
+
+    // Waiters are alerted before callbacks are run
+    waiter_lock.unlock();
     cv_.notify_all();
 
-    // run callbacks, lock not needed since the future is finsihed by this
-    // point so nothing else can modify the callbacks list and it is safe
-    // to iterate.
-    //
-    // In fact, it is important not to hold the locks because the callback
-    // may be slow or do its own locking on other resources
-    for (auto&& callback : callbacks_) {
-      std::move(callback)();
+    std::vector<Callback> callbacks_to_run = std::move(callbacks_);
```

Comment:
@pitrou @bkietz This adds a bit of overhead to callback execution but it ensures that TryAddCallback will only return false if all previously added callbacks have finished.  Callbacks can now be added while the future is in the process of being completed which opens up the door for concurrent access to `callbacks_` hence the lock and move.

Note, this is unrelated to the lockfree queue.  My first iteration of serial readahead used mutex & std::queue and still ran into this problem.

Reply:
Nevermind.  Even with the change it still isn't very safe to try and rely on `AddCallback` ordering.  Parent layers were doing things like...
```
if (fut.IsFinished()) {
  // Do something, this can run before all callbacks run
}
```
Preventing the future itself from being marked finished until after all the callbacks run seems like too much complexity and performance hit.  Instead, if something needs to ensure a callback is run it needs to use `Then` and return a new future even though this added a bit of complexity (need to store shared_ptr in the queue) to serial readback.

Reply:
Also, since this comment is probably confusing now.  It was explaining a change to `Future` that I went back and reverted.  There are no changes to `Future` as part of this PR.